### PR TITLE
feat: NVIDIA NIM on EKS Pattern

### PR DESCRIPTION
**Pull request by hustshawn**
### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

### Motivation

This PR is to address https://github.com/awslabs/data-on-eks/issues/560

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [x] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

**Review by askulkarni2 — COMMENTED** [review]
Can we deploy this using existing Triton + vLLM stack? From NIM docs, it seems like NIM can detect TensorRT-LLM or vLLM profiles for [optimizations](https://docs.nvidia.com/nim/large-language-models/latest/optimization.html).  Also can we include an [observability](https://docs.nvidia.com/nim/large-language-models/latest/observability.html) section?

**Review by vara-bonthu — COMMENTED** [review]
+1 to @askulkarni2 . I had a conversation with @hustshawn  and explained the necessary changes to reuse the existing blueprint, as mentioned in the issue [here](https://github.com/awslabs/data-on-eks/issues/560#issuecomment-2181480657).

1/ Use the existing NVIDIA Triton blueprint.
2/ Create a new file called `nvidia-nim.tf` to add the EFS resources and NIM Helm chart, with a variable called `enable_nvidia_nim=true`.
3/ Add a new variable called `enable_nvidia_triton_server=true` as default, so it can be disabled when NIM is enabled.
4/ Add labels and tolerations to the NVIDIA NIM Helm chart values.
5/ Show observability dashboards using Grafana in the website docs.
6/ Create a python client that reads prompts from a file and generates responses for hosted NIM model.
7/ Create a comparison between NVIDIA Triton with vLLM and NVIDIA NIM using the same model, and publish this as part of the website docs.

**Comment by hustshawn** [comment]
Sure, that makes sense to me. I will make changes accordingly and update you once it's ready for your review again.

**Comment by hustshawn** [comment]
> 7/ Create a comparison between NVIDIA Triton with vLLM and NVIDIA NIM using the same model, and publish this as part of the website docs.

Hi @vara-bonthu , in terms of the comparison, there are some limitations here,
1. current triton_vllm pattern only provides Llama2 and mistral-7b models, while NIM pattern uses Llama3 models;
2. NIM has builtin [vLLM and TRT-LLM](https://docs.nvidia.com/nim/large-language-models/latest/introduction.html#architecture:~:text=NIM%20downloads%20the%20optimized%20TRT%20engine%20and%20runs%20an%20inference%20using%20the%20TRT%2DLLM%20library.%20For%20all%20other%20NVIDIA%20GPUs%2C%20NIM%20downloads%20a%20non%2Doptimized%20model%20and%20runs%20it%20using%20the%20vLLM%20library.), should I just compare Llama3-7b vs Llama3-70b on NIM, so that it can be an apple-apple comparison?
3. I created a similar script as triton_client, but found the result has significant difference. So again I cannot see it's a valid comparison with the existing one mentioned in [triton doc](https://awslabs.github.io/data-on-eks/docs/gen-ai/inference/vLLM-NVIDIATritonServer-Llama2#testing-llama-2-7b-chat-and-mistral-7b-chat-models).

```
Loading inputs from `prompts.txt`...
Model meta/llama3-8b-instruct - Request 14: 4877.96 ms
Model meta/llama3-8b-instruct - Request 10: 6582.67 ms
Model meta/llama3-8b-instruct - Request 3: 7919.11 ms
Model meta/llama3-8b-instruct - Request 15: 7972.63 ms
Model meta/llama3-8b-instruct - Request 1: 8646.89 ms
Model meta/llama3-8b-instruct - Request 5: 8933.86 ms
Model meta/llama3-8b-instruct - Request 12: 9068.71 ms
Model meta/llama3-8b-instruct - Request 18: 9932.78 ms
Model meta/llama3-8b-instruct - Request 0: 10393.83 ms
Model meta/llama3-8b-instruct - Request 6: 10416.12 ms
Model meta/llama3-8b-instruct - Request 16: 10688.25 ms
Model meta/llama3-8b-instruct - Request 4: 10735.83 ms
Model meta/llama3-8b-instruct - Request 11: 10938.80 ms
Model meta/llama3-8b-instruct - Request 8: 11117.52 ms
Model meta/llama3-8b-instruct - Request 17: 12112.03 ms
Model meta/llama3-8b-instruct - Request 2: 12302.77 ms
Model meta/llama3-8b-instruct - Request 19: 12796.39 ms
Model meta/llama3-8b-instruct - Request 9: 13636.33 ms
Model meta/llama3-8b-instruct - Request 13: 13731.89 ms
Model meta/llama3-8b-instruct - Request 7: 15027.77 ms
Storing results into `results.txt`...
Total time for all requests: 207.83 seconds (207832.13 milliseconds)
PASS: NVIDIA NIM example
```

**Comment by hustshawn** [comment]
Hi @vara-bonthu , please review again. Thanks.

**Comment by vara-bonthu** [comment]
Thanks for updating the PR @hustshawn !

@askulkarni2 @ratnopamc please review

**Review by vara-bonthu — COMMENTED** [review]
couple of minor comments for the PR

File: website/docs/gen-ai/inference/nvidia-nim-llama3.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+Model meta/llama3-8b-instruct - Request 6: 10243.54 ms
+Model meta/llama3-8b-instruct - Request 4: 10596.04 ms
+Model meta/llama3-8b-instruct - Request 11: 10798.81 ms
+Model meta/llama3-8b-instruct - Request 17: 10793.39 ms
+Model meta/llama3-8b-instruct - Request 8: 10977.41 ms
+Model meta/llama3-8b-instruct - Request 2: 12085.08 ms
+Model meta/llama3-8b-instruct - Request 19: 12645.12 ms
+Model meta/llama3-8b-instruct - Request 9: 13346.84 ms
+Model meta/llama3-8b-instruct - Request 13: 13581.21 ms
+Model meta/llama3-8b-instruct - Request 7: 14917.04 ms
```
Comment (vara-bonthu) [inline website/docs/gen-ai/inference/nvidia-nim-llama3.md:224]: print seconds as well

Comment (hustshawn) [reply website/docs/gen-ai/inference/nvidia-nim-llama3.md]: done. keep same format as other place (seconds and ms)

File: ai-ml/nvidia-triton-server/variables.tf [hunk 2]
```diff
@@ -47,3 +53,25 @@ variable "huggingface_token" {
   default     = "DUMMY_TOKEN_REPLACE_ME"
   sensitive   = true
 }
+
+variable "enable_nvidia_nim" {
+  description = "Toggle to enable or disable NVIDIA NIM pattern resource creation"
+  default     = true
```
[the remark this answers is not in the document]
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/variables.tf]: done

File: ai-ml/nvidia-triton-server/nvidia-nim.tf [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  chart            = "${path.module}/nim-llm"
+  create_namespace = true
+  namespace        = kubernetes_namespace.nim[count.index].metadata[0].name
+  timeout          = 360
+  wait             = false
+  values = [
+    templatefile(
+      "${path.module}/helm-values/nim-llm.yaml",
+      {
+        ngc_api_key = var.ngc_api_key
```
[the remark this answers is not in the document]
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/nvidia-nim.tf:120]: updated into the install.sh script.

File: ai-ml/nvidia-triton-server/nvidia-nim.tf [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        storage = "100Gi"
+      }
+    }
+  }
+}
+
+#---------------------------------------------------------------
+# NIM LLM Helm Chart
+#---------------------------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/nvidia-nim.tf:80]: done

File: ai-ml/nvidia-triton-server/helm-values/kube-prometheus.yaml [hunk 5]
```diff
@@ -21,28 +21,20 @@ alertmanager:
...
-alertmanager:
-  enabled: false
-
-grafana:
-  enabled: true
-  defaultDashboardsEnabled: true
+  dashboards:
+    default:
+      nim-llm-monitoring:
+        url: https://raw.githubusercontent.com/hustshawn/nim-llm-eks/main/monitoring/nim-llm-monitoring.json
```
[the remark this answers is not in the document]
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/helm-values/kube-prometheus.yaml]: In the latest commit, I have put the json locally and updated reference accordingly.

**Comment by hustshawn** [comment]
@vara-bonthu updated based on our discussion and latest comment. Please review again. Thanks

**Review by vara-bonthu — APPROVED** [review]
LGTM! Thanks for the updates 🙌🏼

**Comment by vara-bonthu** [comment]
@askulkarni2 please review

File: ai-ml/nvidia-triton-server/addons.tf [hunk 6]
```diff
@@ -140,6 +146,20 @@ module "eks_blueprints_addons" {
...
 
+  helm_releases = {
+    "prometheus-adapter" = {
+      repository = "https://prometheus-community.github.io/helm-charts"
+      chart      = "prometheus-adapter"
+      namespace  = module.eks_blueprints_addons.kube_prometheus_stack.namespace
+      version    = "4.10.0"
+      values = [
+        templatefile(
+          "${path.module}/helm-values/prometheus-adapter.yaml", {}
```
Comment (ratnopamc) [inline ai-ml/nvidia-triton-server/addons.tf:157]: Considering that the custom metric for triton-vLLM will be different with nim disabled, we should rename this adapter to something like `prometheus-adpater-nim.yaml`. Idea is only if nim is enabled this particular config will be used.

File: ai-ml/nvidia-triton-server/helm-values/kube-prometheus.yaml [hunk 7]
```diff
@@ -21,28 +21,21 @@ alertmanager:
...
-          resources:
-            requests:
-              storage: 50Gi
-alertmanager:
-  enabled: false
-
-grafana:
-  enabled: true
-  defaultDashboardsEnabled: true
+  dashboards:
```
Comment (ratnopamc) [inline ai-ml/nvidia-triton-server/helm-values/kube-prometheus.yaml:37]: Rename the values file to `kube-prometheus-nim.yaml` and enable nims dashboard. If nims is disabled vanila `kube-prometheus.yaml` should be used.

File: website/docs/gen-ai/inference/nvidia-nim-llama3.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+We are actively enhancing this blueprint to incorporate improvements in observability, logging, and scalability aspects.
+:::
+
+# NVIDIA NIM LLM Deployment on Amazon EKS
+
+## What is NVIDIA NIM?
+
+NVIDIA NIM makes it easy for IT and DevOps teams to self-host large language models (LLMs) in their own managed environments while still providing developers with industry standard APIs that allow them to build powerful copilots, chatbots, and AI assistants that can transform their business. Leveraging NVIDIA’s cutting-edge GPU acceleration and scalable deployment, NIM offers the fastest path to inference with unparalleled performance.
+
```
Comment (ratnopamc) [inline website/docs/gen-ai/inference/nvidia-nim-llama3.md:24]: Add a note or info section here to mention or link to the NIM pricing or licensing information.

**Review by ratnopamc — CHANGES_REQUESTED** [review]
@hustshawn thanks for the PR. Could you please review the comments and address?

File: ai-ml/nvidia-triton-server/addons.tf (hunk 6 above)
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/addons.tf:157]: got your idea. but honestly i don't think it deserve to make it too complex. 
considering there will be cases including 1/enable_triton only; 2. enable_nim only; 3. enable both; then it falls into some difficult cases you have to handle.

current settings/configs dosen't introduce any side effect even if only triton-server enabled; if you have the customized metrics for tritton-vllm later, you can just add those metrics on top of existing one. 
@ratnopamc Does it make sense to you?

File: website/docs/gen-ai/inference/nvidia-nim-llama3.md (hunk 8 above)
Comment (hustshawn) [reply website/docs/gen-ai/inference/nvidia-nim-llama3.md:24]: sure

File: ai-ml/nvidia-triton-server/addons.tf (hunk 6 above)
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/addons.tf]: as aligned offline, will keep the prometheus-adapter addon as it's common for both triton and nim pattern. inject the prometheus namespace into the values with a variable

File: ai-ml/nvidia-triton-server/helm-values/kube-prometheus.yaml (hunk 7 above)
Comment (hustshawn) [reply ai-ml/nvidia-triton-server/helm-values/kube-prometheus.yaml:37]: aligned keep this values file and the nim dashboard won't introduce side effect

**Pull request merged**